[JSC] Fix memory leak in JSContext.name

hyjorc1 opened the pull request:
#### fea636a4940daf25c3850b4f62d2574fbcc2f85e
<pre>
[JSC] Fix memory leak in JSContext.name
<a href="https://bugs.webkit.org/show_bug.cgi?id=267476">https://bugs.webkit.org/show_bug.cgi?id=267476</a>
<a href="https://rdar.apple.com/120665132">rdar://120665132</a>

Reviewed by Yusuke Suzuki.

The caller of `JSGlobalContextCopyName` should be responsible
for managing the memory of the returned `JSStringRef`.

* Source/JavaScriptCore/API/JSContext.mm:
(-[JSContext name]):

Canonical link: <a href="https://commits.webkit.org/273075@main">https://commits.webkit.org/273075@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8ab5576c5c13308cce5c7431d9747a55e5784710

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33946 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12731 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/35901 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36571 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/30759 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15128 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9876 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29811 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34434 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/10770 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30273 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9380 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9490 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30291 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37879 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/28881 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30837 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30610 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35593 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/33861 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9609 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7547 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33490 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11400 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/40412 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7870 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10180 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/8440 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10419 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/API/JSContext.mm:
```diff
@@ -265,7 +265,9 @@ - (NSString *)name
-    return adoptCF(JSStringCopyCFString(kCFAllocatorDefault, name)).bridgingAutorelease();
+    NSString *result = adoptCF(JSStringCopyCFString(kCFAllocatorDefault, name)).bridgingAutorelease();
+    JSStringRelease(name);
+    return result;
```
Is it correct? Now, ownership gets moved from CF to NSString, and it is returning that, right?

hyjorc1 replied:
Yes, this is correct. `.bridgingAutorelease()` will turn it to an auto release object. That means it will be released automatically at a later point in time (typically when the current autorelease pool drains). So, the object in auto release mode will be returned. I have tested it locally with `leaks` and `[result release];`. I believe it's correct.

Constellation replied:
Ah, I see. `name` is leaked. The result is not.

Constellation replied:
How about using `JSRetainPtr`? It makes code safer.

```
auto name = adopt(JSGlobalContextCopyName(m_context));
if (!name)
    return nil;

return adoptCF(JSStringCopyCFString(kCFAllocatorDefault, name.get())).bridgingAutorelease();
```

hyjorc1 replied:
Sweet, even better. Done.

Constellation reviewed (approved):
r=me

hyjorc1 committed:
[JSC] Fix memory leak in JSContext.name
https://bugs.webkit.org/show_bug.cgi?id=267476
rdar://120665132

Reviewed by Yusuke Suzuki.

The caller of `JSGlobalContextCopyName` should be responsible
for managing the memory of the returned `JSStringRef`.

* Source/JavaScriptCore/API/JSContext.mm:
(-[JSContext name]):

Canonical link: https://commits.webkit.org/273075@main
